### ISSUE TITLE
Connection pool issue

### ISSUE THREAD

**Issue by shihadaf** [simoc/csvjdbc#159]
Hi, does the driver support connection pooling?
I tried with a simple connection pool and also with HikariCP, but I got an out of range error in setting the parameters of a prepared statement:
java.sql.SQLException: Parameter index out of range: 1
	at org.relique.jdbc.csv.CsvPreparedStatement.checkParameterIndex(CsvPreparedStatement.java:68)
	at org.relique.jdbc.csv.CsvPreparedStatement.setString(CsvPreparedStatement.java:541)
	
BTW I encounter the error in only 5 (more or less) out of 50 prepared statements, with a connection pool of size 60

**Comment by simoc** [comment]
Thank you for reporting this problem.

I tried CsvJdbc with HikariCP and I also had some Java exceptions using prepared statements.

I think these problems are due to race conditions and/or global variables in CsvJdbc.

My example unit test that demonstrates these problems is:

```
	@Test
	public void testHikariCP() throws SQLException
	{
		HikariConfig config = new HikariConfig();
		config.setJdbcUrl("jdbc:relique:csv:" + filePath);

		try (HikariDataSource ds = new HikariDataSource(config))
		{
			for (int i = 0; i < 10000; i++)
			{
				final int ii = i;
				Thread t = new Thread(() ->
				{
					System.out.println("ii=" + ii);

					try (Connection connection = ds.getConnection())
					{
						PreparedStatement preparedStatement;
						if (ii % 2 == 0)
						{
							preparedStatement = connection.prepareStatement("SELECT * FROM sample WHERE id = ?");
							preparedStatement.setString(1, "A123");
						}
						else
						{
							preparedStatement = connection.prepareStatement("SELECT * FROM sample WHERE id = ? OR id = ?");
							preparedStatement.setString(1, "A123");
							preparedStatement.setString(2, "impossible");
						}
						ResultSet resultSet = preparedStatement.executeQuery();
						assertTrue(resultSet.next());
						assertEquals("A123", resultSet.getString(1));
						assertFalse(resultSet.next());
					}
					catch (SQLException e)
					{
						e.printStackTrace();
					}
				});
				t.start();
			}			
		}
	}
```

Approximately 4% of the iterations failed for me with two types of exceptions:

```
ii=9750
java.sql.SQLException: Parameter index out of range: 1
	at org.relique.jdbc.csv.CsvPreparedStatement.checkParameterIndex(CsvPreparedStatement.java:68)
	at org.relique.jdbc.csv.CsvPreparedStatement.setString(CsvPreparedStatement.java:541)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.setString(HikariProxyPreparedStatement.java)
	at org.relique.jdbc.csv.TestConnectionPool.lambda$0(TestConnectionPool.java:90)
	at java.base/java.lang.Thread.run(Thread.java:834)

[...]

ii=9770
Exception in thread "Thread-9770" org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
	at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
	at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:31)
	at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:183)
	at org.relique.jdbc.csv.TestConnectionPool.lambda$0(TestConnectionPool.java:94)
	at java.base/java.lang.Thread.run(Thread.java:834)
``

**Comment by simoc** [comment]
The problem is the global prepared statement position counter Placeholder.nextIndex, that counts each '?' that occurs in the prepared statement.

https://github.com/simoc/csvjdbc/blob/5c99fcf8317478250b70a9d36ee213ececc45fa2/src/main/java/org/relique/jdbc/csv/Placeholder.java#L30

When two threads parse SQL prepared statements at the same time, this counter goes wrong, and a `java.sql.SQLException: Parameter index out of range: 1` is thrown.

After I refactored this class to remove this global counter, there is no exception from HikariCP. I just need to decide which is the cleanest implementation for this for change.

**Comment by shihadaf** [comment]
Thanks for the quick solution, appreciated!

**Comment by simoc** [comment]
This problem is fixed in Csvjdbc version 1.0.46.

Download it from https://sourceforge.net/projects/csvjdbc/files/ or from https://central.sonatype.com/
